### ISSUE TITLE
Math repo runs no tests on Windows

### ISSUE THREAD

**Issue by seantalts** [stan-dev/math#951]
Ideally we'd run the unit tests there. I think they might be broken, though.

#### Current Version:
v2.18.0

**Comment by syclik** [comment]
Thanks. I've created a branch: feature/issue-951-windows-ci-tests

If anyone wants to help, much appreciated.

@seantalts, I don't have direct access to a windows box anymore. Can I log
into gelman-group-win periodically to fix this?

**Comment by syclik** [comment]
Thanks! I think I still have credentials. I'll let you know if I don't.

**Comment by rok-cesnovar** [comment]
I am not sure if this helps in any way, my knowledge of CI and all that is negligible :) I hope it does.

I did a clean clone of the math repo on my windows machine. The tests run when CC=g++ but fail with clang++

```
clang++ -Wall -I . -isystem lib/eigen_3.3.3 -isystem lib/boost_1.66.0 -isystem lib/sundials_3.1.0/include -std=c++1y -DBOOST_RESULT_OF_USE_TR1 -DBOOST_NO_DECLTYPE -DBOOST_DISABLE_ASSERTS -DBOOST_PHOENIX_NO_VARIADIC_EXPRESSION -Wno-unused-function -Wno-uninitialized -DNO_FPRINTF_OUTPUT -pipe   -shared -fPIC -o test/unit/libmultiple.so test/unit/multiple_translation_units1.o test/unit/multiple_translation_units2.o
test/unit/multiple_translation_units2.o: file not recognized: File truncated
clang++.exe: error: linker command failed with exit code 1 (use -v to see invocation)
make: *** [test/unit/libmultiple.so] Error 1
```

**Comment by bob-carpenter** [comment]
> On Jul 24, 2018, at 11:45 PM, Rok Češnovar <notifications@github.com> wrote:
> 
> I am not sure if this helps in any way, my knowledge of CI and all that is negligible :) I hope it does.
> 
> I did a clean clone of the math repo on my windows machine. The tests run when CC=g++

Great news.

> but fail with clang++
> 
> clang++ -Wall -I . -isystem lib/eigen_3.3.3 -isystem lib/boost_1.66.0 -isystem lib/sundials_3.1.0/include -std=c++1y -DBOOST_RESULT_OF_USE_TR1 -DBOOST_NO_DECLTYPE -DBOOST_DISABLE_ASSERTS -DBOOST_PHOENIX_NO_VARIADIC_EXPRESSION -Wno-unused-function -Wno-uninitialized -DNO_FPRINTF_OUTPUT -pipe   -shared -fPIC -o test/unit/libmultiple.so test/unit/multiple_translation_units1.o test/unit/multiple_translation_units2.o
> test/unit/multiple_translation_units2.o: file not recognized: File truncated
> clang++.exe: error: linker command failed with exit code 1 (use -v to see invocation)
> make: *** [test/unit/libmultiple.so] Error 1

We used to run into lots of problems in Windows with the assembler segfaulting.  It was usually due to too-large translation units.  Setting the optimizationlevel above 0 and turning off debug can help---I don't see any of those flags and I'm not sure what the defaults are.

**Comment by rok-cesnovar** [comment]
The OS is Windows 10 Enterprise x64, the version of clang is 3.7. That is probably a bit too old of a version? I will check if upgrading solves anything.

**Comment by syclik** [comment]
I don't know. I've never successfully installed clang on Windows!

**Comment by syclik** [comment]
I think we'll need to fix #954 first.

**Comment by rok-cesnovar** [comment]
I went back and checked a few older versions of stan math.

The assign_test.cpp breaks after [this](https://github.com/stan-dev/math/commit/1f0e07bbff532c88452200f500a0ad92f8c08568) commit.

**Comment by syclik** [comment]
I think I found the fundamental issue with the assign test that @rok-cesnovar narrowed down. (Thank you, btw.)

`size_t` in Windows on that compiler with the C++11 flag is set to `long long unsigned int`. We don't have that specialization in `stan/math/prim/scal/meta/ad_promotable.hpp`.

If I add that specialization, that test passes. (I tried the cheap thing of just adding `size_t` to `ad_promotable`, but got a redefinition on my Mac.) Since we're now C++11, I think we need to add the rest of these types on this page and perhaps that'll solve enough of our problems:
https://en.cppreference.com/w/cpp/language/types

@bob-carpenter, any thoughts? Is that the right thing to do?

**Comment by syclik** [comment]
The next one that fails is this: 
```
runTests.py test/unit/math/mix/mat/fun/mean_test.cpp
```

The error message is:
```
>runTests.py test/unit/math/mix/mat/fun/mean_test.cpp
------------------------------------------------------------
make -j1 test/unit/math/mix/mat/fun/mean_test.exe
g++ -std=c++1y  -m64 -Wall -Wno-unused-function -Wno-uninitialized -DBOOST_RESULT_OF_USE_TR1 -DBOOST_NO_DECLTYPE -DBOOST_DISABLE_ASSERTS -DBOOST_PHOENIX_NO_VARIADIC_EXPRESSION  -DNO_FPRINTF_OUTPUT -pipe   -O3 -I . -I lib/eigen_3.3.3 -I lib/boost_1.66.0 -I lib/sundials_3.1.0/include -DGTEST_USE_OWN_TR1_TUPLE -DGTEST_HAS_PTHREAD=0 -I lib/gtest_1.7.0/include -I lib/gtest_1.7.0   -c -o test/unit/math/mix/mat/fun/mean_test.o test/unit/math/mix/mat/fun/mean_test.cpp
In file included from lib/eigen_3.3.3/Eigen/Core:457:0,
                 from lib/eigen_3.3.3/Eigen/Dense:1,
                 from ./stan/math/prim/mat/fun/Eigen.hpp:4,
                 from ./stan/math/mix/mat/fun/typedefs.hpp:4,
                 from ./stan/math/mix/mat.hpp:4,
                 from test/unit/math/mix/mat/fun/mean_test.cpp:1:
lib/eigen_3.3.3/Eigen/src/Core/Redux.h: In instantiation of 'typename Eigen::internal::traits<T>::Scalar Eigen::DenseBase<Derived>::mean() const [with Derived = Eigen::Matrix<stan::math::fvar<stan::math::fvar<stan::math::var> >, -1, 1>; typename Eigen::internal::traits<T>::Scalar = stan::math::fvar<stan::math::fvar<stan::math::var> >]':
./stan/math/prim/mat/fun/mean.hpp:40:17:   required from 'typename boost::math::tools::promote_args<T>::type stan::math::mean(const Eigen::Matrix<T, R, C>&) [with T = stan::math::fvar<stan::math::fvar<stan::math::var> >; int R = -1; int C = 1; typename boost::math::tools::promote_args<T>::type = stan::math::fvar<stan::math::fvar<stan::math::var> >]'
test/unit/math/mix/mat/fun/mean_test.cpp:287:19:   required from here
lib/eigen_3.3.3/Eigen/src/Core/Redux.h:468:83: error: no matching function for call to 'stan::math::fvar<stan::math::fvar<stan::math::var> >::fvar(Eigen::EigenBase<Eigen::Matrix<stan::math::fvar<stan::math::fvar<stan::math::var> >, -1, 1> >::Index)'
   return Scalar(derived().redux(Eigen::internal::scalar_sum_op<Scalar,Scalar>())) / Scalar(this->size());
                                                                                   ^
lib/eigen_3.3.3/Eigen/src/Core/Redux.h:468:83: note: candidates are:
In file included from ./stan/math/fwd/core.hpp:4:0,
                 from ./stan/math/mix/mat/fun/typedefs.hpp:5,
                 from ./stan/math/mix/mat.hpp:4,
                 from test/unit/math/mix/mat/fun/mean_test.cpp:1:
./stan/math/fwd/core/fvar.hpp:123:3: note: template<class V, class D> stan::math::fvar<T>::fvar(const V&, const D&)
   fvar(const V& v, const D& d) : val_(v), d_(d) {
   ^
./stan/math/fwd/core/fvar.hpp:123:3: note:   template argument deduction/substitution failed:
In file included from lib/eigen_3.3.3/Eigen/Core:457:0,
                 from lib/eigen_3.3.3/Eigen/Dense:1,
                 from ./stan/math/prim/mat/fun/Eigen.hpp:4,
                 from ./stan/math/mix/mat/fun/typedefs.hpp:4,
                 from ./stan/math/mix/mat.hpp:4,
                 from test/unit/math/mix/mat/fun/mean_test.cpp:1:
lib/eigen_3.3.3/Eigen/src/Core/Redux.h:468:83: note:   candidate expects 2 arguments, 1 provided
   return Scalar(derived().redux(Eigen::internal::scalar_sum_op<Scalar,Scalar>())) / Scalar(this->size());
                                                                                   ^
In file included from ./stan/math/fwd/core.hpp:4:0,
                 from ./stan/math/mix/mat/fun/typedefs.hpp:5,
                 from ./stan/math/mix/mat.hpp:4,
                 from test/unit/math/mix/mat/fun/mean_test.cpp:1:
./stan/math/fwd/core/fvar.hpp:104:3: note: template<class V> stan::math::fvar<T>::fvar(const V&, typename boost::enable_if_c<stan::math::ad_promotable<V, T>::value>::type*)
   fvar(const V& v,
   ^
./stan/math/fwd/core/fvar.hpp:104:3: note:   template argument deduction/substitution failed:
./stan/math/fwd/core/fvar.hpp: In substitution of 'template<class V> stan::math::fvar<T>::fvar(const V&, typename boost::enable_if_c<stan::math::ad_promotable<V, T>::value>::type*) [with V = long long int]':
lib/eigen_3.3.3/Eigen/src/Core/Redux.h:468:83:   required from 'typename Eigen::internal::traits<T>::Scalar Eigen::DenseBase<Derived>::mean() const [with Derived = Eigen::Matrix<stan::math::fvar<stan::math::fvar<stan::math::var> >, -1, 1>; typename Eigen::internal::traits<T>::Scalar = stan::math::fvar<stan::math::fvar<stan::math::var> >]'
./stan/math/prim/mat/fun/mean.hpp:40:17:   required from 'typename boost::math::tools::promote_args<T>::type stan::math::mean(const Eigen::Matrix<T, R, C>&) [with T = stan::math::fvar<stan::math::fvar<stan::math::var> >; int R = -1; int C = 1; typename boost::math::tools::promote_args<T>::type = stan::math::fvar<stan::math::fvar<stan::math::var> >]'
test/unit/math/mix/mat/fun/mean_test.cpp:287:19:   required from here
./stan/math/fwd/core/fvar.hpp:104:3: error: no type named 'type' in 'struct boost::enable_if_c<false, void>'
lib/eigen_3.3.3/Eigen/src/Core/Redux.h: In instantiation of 'typename Eigen::internal::traits<T>::Scalar Eigen::DenseBase<Derived>::mean() const [with Derived = Eigen::Matrix<stan::math::fvar<stan::math::fvar<stan::math::var> >, -1, 1>; typename Eigen::internal::traits<T>::Scalar = stan::math::fvar<stan::math::fvar<stan::math::var> >]':
./stan/math/prim/mat/fun/mean.hpp:40:17:   required from 'typename boost::math::tools::promote_args<T>::type stan::math::mean(const Eigen::Matrix<T, R, C>&) [with T = stan::math::fvar<stan::math::fvar<stan::math::var> >; int R = -1; int C = 1; typename boost::math::tools::promote_args<T>::type = stan::math::fvar<stan::math::fvar<stan::math::var> >]'
test/unit/math/mix/mat/fun/mean_test.cpp:287:19:   required from here
./stan/math/fwd/core/fvar.hpp:88:3: note: stan::math::fvar<T>::fvar(const T&) [with T = stan::math::fvar<stan::math::var>]
   fvar(const T& v) : val_(v), d_(0.0) {  // NOLINT(runtime/explicit)
   ^
./stan/math/fwd/core/fvar.hpp:88:3: note:   no known conversion for argument 1 from 'Eigen::EigenBase<Eigen::Matrix<stan::math::fvar<stan::math::fvar<stan::math::var> >, -1, 1> >::Index {aka long long int}' to 'const stan::math::fvar<stan::math::var>&'
./stan/math/fwd/core/fvar.hpp:78:3: note: stan::math::fvar<T>::fvar(const stan::math::fvar<T>&) [with T = stan::math::fvar<stan::math::var>]
   fvar(const fvar<T>& x) : val_(x.val_), d_(x.d_) {}
   ^
./stan/math/fwd/core/fvar.hpp:78:3: note:   no known conversion for argument 1 from 'Eigen::EigenBase<Eigen::Matrix<stan::math::fvar<stan::math::fvar<stan::math::var> >, -1, 1> >::Index {aka long long int}' to 'const stan::math::fvar<stan::math::fvar<stan::math::var> >&'
./stan/math/fwd/core/fvar.hpp:70:3: note: stan::math::fvar<T>::fvar() [with T = stan::math::fvar<stan::math::var>]
   fvar() : val_(0.0), d_(0.0) {}
   ^
./stan/math/fwd/core/fvar.hpp:70:3: note:   candidate expects 0 arguments, 1 provided
In file included from lib/eigen_3.3.3/Eigen/Core:457:0,
                 from lib/eigen_3.3.3/Eigen/Dense:1,
                 from ./stan/math/prim/mat/fun/Eigen.hpp:4,
                 from ./stan/math/mix/mat/fun/typedefs.hpp:4,
                 from ./stan/math/mix/mat.hpp:4,
                 from test/unit/math/mix/mat/fun/mean_test.cpp:1:
lib/eigen_3.3.3/Eigen/src/Core/Redux.h: In instantiation of 'typename Eigen::internal::traits<T>::Scalar Eigen::DenseBase<Derived>::mean() const [with Derived = Eigen::Matrix<stan::math::fvar<stan::math::fvar<stan::math::var> >, 1, -1>; typename Eigen::internal::traits<T>::Scalar = stan::math::fvar<stan::math::fvar<stan::math::var> >]':
./stan/math/prim/mat/fun/mean.hpp:40:17:   required from 'typename boost::math::tools::promote_args<T>::type stan::math::mean(const Eigen::Matrix<T, R, C>&) [with T = stan::math::fvar<stan::math::fvar<stan::math::var> >; int R = 1; int C = -1; typename boost::math::tools::promote_args<T>::type = stan::math::fvar<stan::math::fvar<stan::math::var> >]'
test/unit/math/mix/mat/fun/mean_test.cpp:410:19:   required from here
lib/eigen_3.3.3/Eigen/src/Core/Redux.h:468:83: error: no matching function for call to 'stan::math::fvar<stan::math::fvar<stan::math::var> >::fvar(Eigen::EigenBase<Eigen::Matrix<stan::math::fvar<stan::math::fvar<stan::math::var> >, 1, -1> >::Index)'
   return Scalar(derived().redux(Eigen::internal::scalar_sum_op<Scalar,Scalar>())) / Scalar(this->size());
                                                                                   ^
lib/eigen_3.3.3/Eigen/src/Core/Redux.h:468:83: note: candidates are:
In file included from ./stan/math/fwd/core.hpp:4:0,
                 from ./stan/math/mix/mat/fun/typedefs.hpp:5,
                 from ./stan/math/mix/mat.hpp:4,
                 from test/unit/math/mix/mat/fun/mean_test.cpp:1:
./stan/math/fwd/core/fvar.hpp:123:3: note: template<class V, class D> stan::math::fvar<T>::fvar(const V&, const D&)
   fvar(const V& v, const D& d) : val_(v), d_(d) {
   ^
./stan/math/fwd/core/fvar.hpp:123:3: note:   template argument deduction/substitution failed:
In file included from lib/eigen_3.3.3/Eigen/Core:457:0,
                 from lib/eigen_3.3.3/Eigen/Dense:1,
                 from ./stan/math/prim/mat/fun/Eigen.hpp:4,
                 from ./stan/math/mix/mat/fun/typedefs.hpp:4,
                 from ./stan/math/mix/mat.hpp:4,
                 from test/unit/math/mix/mat/fun/mean_test.cpp:1:
lib/eigen_3.3.3/Eigen/src/Core/Redux.h:468:83: note:   candidate expects 2 arguments, 1 provided
   return Scalar(derived().redux(Eigen::internal::scalar_sum_op<Scalar,Scalar>())) / Scalar(this->size());
                                                                                   ^
In file included from ./stan/math/fwd/core.hpp:4:0,
                 from ./stan/math/mix/mat/fun/typedefs.hpp:5,
                 from ./stan/math/mix/mat.hpp:4,
                 from test/unit/math/mix/mat/fun/mean_test.cpp:1:
./stan/math/fwd/core/fvar.hpp:104:3: note: template<class V> stan::math::fvar<T>::fvar(const V&, typename boost::enable_if_c<stan::math::ad_promotable<V, T>::value>::type*)
   fvar(const V& v,
   ^
./stan/math/fwd/core/fvar.hpp:104:3: note:   template argument deduction/substitution failed:
./stan/math/fwd/core/fvar.hpp: In substitution of 'template<class V> stan::math::fvar<T>::fvar(const V&, typename boost::enable_if_c<stan::math::ad_promotable<V, T>::value>::type*) [with V = long long int]':
lib/eigen_3.3.3/Eigen/src/Core/Redux.h:468:83:   required from 'typename Eigen::internal::traits<T>::Scalar Eigen::DenseBase<Derived>::mean() const [with Derived = Eigen::Matrix<stan::math::fvar<stan::math::fvar<stan::math::var> >, 1, -1>; typename Eigen::internal::traits<T>::Scalar = stan::math::fvar<stan::math::fvar<stan::math::var> >]'
./stan/math/prim/mat/fun/mean.hpp:40:17:   required from 'typename boost::math::tools::promote_args<T>::type stan::math::mean(const Eigen::Matrix<T, R, C>&) [with T = stan::math::fvar<stan::math::fvar<stan::math::var> >; int R = 1; int C = -1; typename boost::math::tools::promote_args<T>::type = stan::math::fvar<stan::math::fvar<stan::math::var> >]'
test/unit/math/mix/mat/fun/mean_test.cpp:410:19:   required from here
./stan/math/fwd/core/fvar.hpp:104:3: error: no type named 'type' in 'struct boost::enable_if_c<false, void>'
lib/eigen_3.3.3/Eigen/src/Core/Redux.h: In instantiation of 'typename Eigen::internal::traits<T>::Scalar Eigen::DenseBase<Derived>::mean() const [with Derived = Eigen::Matrix<stan::math::fvar<stan::math::fvar<stan::math::var> >, 1, -1>; typename Eigen::internal::traits<T>::Scalar = stan::math::fvar<stan::math::fvar<stan::math::var> >]':
./stan/math/prim/mat/fun/mean.hpp:40:17:   required from 'typename boost::math::tools::promote_args<T>::type stan::math::mean(const Eigen::Matrix<T, R, C>&) [with T = stan::math::fvar<stan::math::fvar<stan::math::var> >; int R = 1; int C = -1; typename boost::math::tools::promote_args<T>::type = stan::math::fvar<stan::math::fvar<stan::math::var> >]'
test/unit/math/mix/mat/fun/mean_test.cpp:410:19:   required from here
./stan/math/fwd/core/fvar.hpp:88:3: note: stan::math::fvar<T>::fvar(const T&) [with T = stan::math::fvar<stan::math::var>]
   fvar(const T& v) : val_(v), d_(0.0) {  // NOLINT(runtime/explicit)
   ^
./stan/math/fwd/core/fvar.hpp:88:3: note:   no known conversion for argument 1 from 'Eigen::EigenBase<Eigen::Matrix<stan::math::fvar<stan::math::fvar<stan::math::var> >, 1, -1> >::Index {aka long long int}' to 'const stan::math::fvar<stan::math::var>&'
./stan/math/fwd/core/fvar.hpp:78:3: note: stan::math::fvar<T>::fvar(const stan::math::fvar<T>&) [with T = stan::math::fvar<stan::math::var>]
   fvar(const fvar<T>& x) : val_(x.val_), d_(x.d_) {}
   ^
./stan/math/fwd/core/fvar.hpp:78:3: note:   no known conversion for argument 1 from 'Eigen::EigenBase<Eigen::Matrix<stan::math::fvar<stan::math::fvar<stan::math::var> >, 1, -1> >::Index {aka long long int}' to 'const stan::math::fvar<stan::math::fvar<stan::math::var> >&'
./stan/math/fwd/core/fvar.hpp:70:3: note: stan::math::fvar<T>::fvar() [with T = stan::math::fvar<stan::math::var>]
   fvar() : val_(0.0), d_(0.0) {}
   ^
./stan/math/fwd/core/fvar.hpp:70:3: note:   candidate expects 0 arguments, 1 provided
In file included from lib/eigen_3.3.3/Eigen/Core:457:0,
                 from lib/eigen_3.3.3/Eigen/Dense:1,
                 from ./stan/math/prim/mat/fun/Eigen.hpp:4,
                 from ./stan/math/mix/mat/fun/typedefs.hpp:4,
                 from ./stan/math/mix/mat.hpp:4,
                 from test/unit/math/mix/mat/fun/mean_test.cpp:1:
lib/eigen_3.3.3/Eigen/src/Core/Redux.h: In instantiation of 'typename Eigen::internal::traits<T>::Scalar Eigen::DenseBase<Derived>::mean() const [with Derived = Eigen::Matrix<stan::math::fvar<stan::math::fvar<stan::math::var> >, -1, -1>; typename Eigen::internal::traits<T>::Scalar = stan::math::fvar<stan::math::fvar<stan::math::var> >]':
./stan/math/prim/mat/fun/mean.hpp:40:17:   required from 'typename boost::math::tools::promote_args<T>::type stan::math::mean(const Eigen::Matrix<T, R, C>&) [with T = stan::math::fvar<stan::math::fvar<stan::math::var> >; int R = -1; int C = -1; typename boost::math::tools::promote_args<T>::type = stan::math::fvar<stan::math::fvar<stan::math::var> >]'
test/unit/math/mix/mat/fun/mean_test.cpp:533:19:   required from here
lib/eigen_3.3.3/Eigen/src/Core/Redux.h:468:83: error: no matching function for call to 'stan::math::fvar<stan::math::fvar<stan::math::var> >::fvar(Eigen::EigenBase<Eigen::Matrix<stan::math::fvar<stan::math::fvar<stan::math::var> >, -1, -1> >::Index)'
   return Scalar(derived().redux(Eigen::internal::scalar_sum_op<Scalar,Scalar>())) / Scalar(this->size());
                                                                                   ^
lib/eigen_3.3.3/Eigen/src/Core/Redux.h:468:83: note: candidates are:
In file included from ./stan/math/fwd/core.hpp:4:0,
                 from ./stan/math/mix/mat/fun/typedefs.hpp:5,
                 from ./stan/math/mix/mat.hpp:4,
                 from test/unit/math/mix/mat/fun/mean_test.cpp:1:
./stan/math/fwd/core/fvar.hpp:123:3: note: template<class V, class D> stan::math::fvar<T>::fvar(const V&, const D&)
   fvar(const V& v, const D& d) : val_(v), d_(d) {
   ^
./stan/math/fwd/core/fvar.hpp:123:3: note:   template argument deduction/substitution failed:
In file included from lib/eigen_3.3.3/Eigen/Core:457:0,
                 from lib/eigen_3.3.3/Eigen/Dense:1,
                 from ./stan/math/prim/mat/fun/Eigen.hpp:4,
                 from ./stan/math/mix/mat/fun/typedefs.hpp:4,
                 from ./stan/math/mix/mat.hpp:4,
                 from test/unit/math/mix/mat/fun/mean_test.cpp:1:
lib/eigen_3.3.3/Eigen/src/Core/Redux.h:468:83: note:   candidate expects 2 arguments, 1 provided
   return Scalar(derived().redux(Eigen::internal::scalar_sum_op<Scalar,Scalar>())) / Scalar(this->size());
                                                                                   ^
In file included from ./stan/math/fwd/core.hpp:4:0,
                 from ./stan/math/mix/mat/fun/typedefs.hpp:5,
                 from ./stan/math/mix/mat.hpp:4,
                 from test/unit/math/mix/mat/fun/mean_test.cpp:1:
./stan/math/fwd/core/fvar.hpp:104:3: note: template<class V> stan::math::fvar<T>::fvar(const V&, typename boost::enable_if_c<stan::math::ad_promotable<V, T>::value>::type*)
   fvar(const V& v,
   ^
./stan/math/fwd/core/fvar.hpp:104:3: note:   template argument deduction/substitution failed:
./stan/math/fwd/core/fvar.hpp: In substitution of 'template<class V> stan::math::fvar<T>::fvar(const V&, typename boost::enable_if_c<stan::math::ad_promotable<V, T>::value>::type*) [with V = long long int]':
lib/eigen_3.3.3/Eigen/src/Core/Redux.h:468:83:   required from 'typename Eigen::internal::traits<T>::Scalar Eigen::DenseBase<Derived>::mean() const [with Derived = Eigen::Matrix<stan::math::fvar<stan::math::fvar<stan::math::var> >, -1, -1>; typename Eigen::internal::traits<T>::Scalar = stan::math::fvar<stan::math::fvar<stan::math::var> >]'
./stan/math/prim/mat/fun/mean.hpp:40:17:   required from 'typename boost::math::tools::promote_args<T>::type stan::math::mean(const Eigen::Matrix<T, R, C>&) [with T = stan::math::fvar<stan::math::fvar<stan::math::var> >; int R = -1; int C = -1; typename boost::math::tools::promote_args<T>::type = stan::math::fvar<stan::math::fvar<stan::math::var> >]'
test/unit/math/mix/mat/fun/mean_test.cpp:533:19:   required from here
./stan/math/fwd/core/fvar.hpp:104:3: error: no type named 'type' in 'struct boost::enable_if_c<false, void>'
lib/eigen_3.3.3/Eigen/src/Core/Redux.h: In instantiation of 'typename Eigen::internal::traits<T>::Scalar Eigen::DenseBase<Derived>::mean() const [with Derived = Eigen::Matrix<stan::math::fvar<stan::math::fvar<stan::math::var> >, -1, -1>; typename Eigen::internal::traits<T>::Scalar = stan::math::fvar<stan::math::fvar<stan::math::var> >]':
./stan/math/prim/mat/fun/mean.hpp:40:17:   required from 'typename boost::math::tools::promote_args<T>::type stan::math::mean(const Eigen::Matrix<T, R, C>&) [with T = stan::math::fvar<stan::math::fvar<stan::math::var> >; int R = -1; int C = -1; typename boost::math::tools::promote_args<T>::type = stan::math::fvar<stan::math::fvar<stan::math::var> >]'
test/unit/math/mix/mat/fun/mean_test.cpp:533:19:   required from here
./stan/math/fwd/core/fvar.hpp:88:3: note: stan::math::fvar<T>::fvar(const T&) [with T = stan::math::fvar<stan::math::var>]
   fvar(const T& v) : val_(v), d_(0.0) {  // NOLINT(runtime/explicit)
   ^
./stan/math/fwd/core/fvar.hpp:88:3: note:   no known conversion for argument 1 from 'Eigen::EigenBase<Eigen::Matrix<stan::math::fvar<stan::math::fvar<stan::math::var> >, -1, -1> >::Index {aka long long int}' to 'const stan::math::fvar<stan::math::var>&'
./stan/math/fwd/core/fvar.hpp:78:3: note: stan::math::fvar<T>::fvar(const stan::math::fvar<T>&) [with T = stan::math::fvar<stan::math::var>]
   fvar(const fvar<T>& x) : val_(x.val_), d_(x.d_) {}
   ^
./stan/math/fwd/core/fvar.hpp:78:3: note:   no known conversion for argument 1 from 'Eigen::EigenBase<Eigen::Matrix<stan::math::fvar<stan::math::fvar<stan::math::var> >, -1, -1> >::Index {aka long long int}' to 'const stan::math::fvar<stan::math::fvar<stan::math::var> >&'
./stan/math/fwd/core/fvar.hpp:70:3: note: stan::math::fvar<T>::fvar() [with T = stan::math::fvar<stan::math::var>]
   fvar() : val_(0.0), d_(0.0) {}
   ^
./stan/math/fwd/core/fvar.hpp:70:3: note:   candidate expects 0 arguments, 1 provided
In file included from lib/eigen_3.3.3/Eigen/Core:457:0,
                 from lib/eigen_3.3.3/Eigen/Dense:1,
                 from ./stan/math/prim/mat/fun/Eigen.hpp:4,
                 from ./stan/math/mix/mat/fun/typedefs.hpp:4,
                 from ./stan/math/mix/mat.hpp:4,
                 from test/unit/math/mix/mat/fun/mean_test.cpp:1:
lib/eigen_3.3.3/Eigen/src/Core/Redux.h: In member function 'typename Eigen::internal::traits<T>::Scalar Eigen::DenseBase<Derived>::mean() const [with Derived = Eigen::Matrix<stan::math::fvar<stan::math::fvar<stan::math::var> >, -1, 1>; typename Eigen::internal::traits<T>::Scalar = stan::math::fvar<stan::math::fvar<stan::math::var> >]':
lib/eigen_3.3.3/Eigen/src/Core/Redux.h:472:1: warning: control reaches end of non-void function [-Wreturn-type]
 }
 ^
lib/eigen_3.3.3/Eigen/src/Core/Redux.h: In member function 'typename Eigen::internal::traits<T>::Scalar Eigen::DenseBase<Derived>::mean() const [with Derived = Eigen::Matrix<stan::math::fvar<stan::math::fvar<stan::math::var> >, 1, -1>; typename Eigen::internal::traits<T>::Scalar = stan::math::fvar<stan::math::fvar<stan::math::var> >]':
lib/eigen_3.3.3/Eigen/src/Core/Redux.h:472:1: warning: control reaches end of non-void function [-Wreturn-type]
 }
 ^
lib/eigen_3.3.3/Eigen/src/Core/Redux.h: In member function 'typename Eigen::internal::traits<T>::Scalar Eigen::DenseBase<Derived>::mean() const [with Derived = Eigen::Matrix<stan::math::fvar<stan::math::fvar<stan::math::var> >, -1, -1>; typename Eigen::internal::traits<T>::Scalar = stan::math::fvar<stan::math::fvar<stan::math::var> >]':
lib/eigen_3.3.3/Eigen/src/Core/Redux.h:472:1: warning: control reaches end of non-void function [-Wreturn-type]
 }
 ^
make: *** [test/unit/math/mix/mat/fun/mean_test.o] Error 1
make -j1 test/unit/math/mix/mat/fun/mean_test.exe failed
exit now (07/26/18 23:36:58 Eastern Daylight Time)
```

**Comment by rok-cesnovar** [comment]
So far what I was only able to figure out is that replacing `m.mean()` with `m.sum()/m.size()` [here](https://github.com/stan-dev/math/blob/develop/stan/math/prim/mat/fun/mean.hpp#L40) fixes this problem and this test passes.

**Comment by syclik** [comment]
I bet this is a similar problem as the assign test. I bet we don't have a specialization for `long long` and so the compile fails. 

I don't know what's different about `fvar<fvar<var>>` that would make this fail, other than it might have to do with how we instantiated the Eigen traits. Actually, that might be it. We should look at the instantiation of the traits for mixed mode to make sure it lines up with reverse and forward mode. (I'm just leaving notes for myself for later.)

**Comment by bob-carpenter** [comment]
The fvar code is templated rather than specialized for type.

Something underneath might not handle long long, but it's not the core of fvar.

A likely culprit is rev, which does not have a specialization for long long.

The only way to get around having all these template definitions is to have a template with specified instantations as we do for the grammars.  Otherwise, there won't be enough symbols defined in the object file to link.

Now that we allow C++11, we could add a long long specialization to the var() constructor.

**Comment by rok-cesnovar** [comment]
Hopefully this info is of any use. 

Other then a few tests that fail for the same `long long` (size_t) problems as the assign_test. these are the remaining tests in `math/mix` that fail to compile on Windows:

- in mix/mat/fun/singular_values_test.cpp there are a few errors like this 
```
test/unit/math/mix/mat/fun/singular_values_test.cpp:72:27:   required from here
lib/eigen_3.3.3/Eigen/src/QR/ColPivHouseholderQR.h:509:116: error: no matching function for call to 'stan::math::fvar<stan::math::fvar<stan::math::var> >::fvar(Eigen::Index&)'
   RealScalar threshold_helper =  numext::abs2<Scalar>(m_colNormsUpdated.maxCoeff() * NumTraits<Scalar>::epsilon()) / RealScalar(rows);
```
Full error message:
[singular_values_test.txt](https://github.com/stan-dev/math/files/2236893/singular_values_test.txt)

- in mix/mat/prob/multi_normal_cholesky_test.cpp
```
In file included from ./stan/math/prim/mat/prob/multi_normal_cholesky_log.hpp:5:0,
                 from ./stan/math/prim/mat.hpp:296,
                 from ./stan/math/rev/mat.hpp:12,
                 from ./stan/math/mix/mat.hpp:6,
                 from test/unit/math/mix/mat/prob/multi_normal_cholesky_test.cpp:1:
./stan/math/prim/mat/prob/multi_normal_cholesky_lpdf.hpp:152:49: note:   cannot convert '((const Eigen::DenseBase<Eigen::Matrix<stan::math::fvar<stan::math::var>, -1, -1> >*)(& inv_L_dbl))->Eigen::DenseBase<Derived>::transpose<Eigen::Matrix<stan::math::fvar<stan::math::var>, -1, -1> >()' (type 'Eigen::DenseBase<Eigen::Matrix<stan::math::fvar<stan::math::var>, -1, -1> >::ConstTransposeReturnType {aka const Eigen::Transpose<const Eigen::Matrix<stan::math::fvar<stan::math::var>, -1, -1> >}') to type 'const StorageBaseType& {aka const Eigen::MatrixBase<Eigen::Transpose<Eigen::Map<Eigen::Matrix<double, -1, 1>, 0, Eigen::Stride<0, 0> > > >&}'
       ops_partials.edge3_.partials_ -= size_vec * inv_L_dbl.transpose();
```
Full error message:
[multi_normal_cholesky_test.txt](https://github.com/stan-dev/math/files/2237006/multi_normal_cholesky_test.txt)

**Comment by rok-cesnovar** [comment]
There is another one in `math/prim/scal/fun/log_modified_bessel_first_kind_test.cpp` that is caused due to M_PI not being standard C++ ([link](https://github.com/robotology/QA/issues/81), [link](https://stackoverflow.com/questions/26065359/m-pi-flagged-as-undeclared-identifier))

```
test/unit/math/prim/scal/fun/log_modified_bessel_first_kind_test.cpp:17:33: error: 'M_PI' was not declared in this scope
                   log(sqrt(2 / (M_PI * 10)) * std::sinh(10)));
```

**Comment by bob-carpenter** [comment]
@rok-cesnovar could you open an issue and fix the problem with `M_PI` and related constants?  

Is the issue that we should be using functions instead of constants here?

**Comment by syclik** [comment]
@rok-cesnovar is ahead! Here's the open issue for it: #986. The fix is on
the makefile branch.

We should have followed up on this issue with the new issue.

**Comment by syclik** [comment]
This was fixed with #1046.
